xlators/cluster/afr/src/afr-open.c

```diff
@@ -271,34 +294,194 @@ afr_fd_ctx_need_open(fd_t *fd, xlator_t *this, unsigned char *need_open)
+    afr_private_t *priv = NULL;
+    int i = 0;
+    int need_open_count = 0;
+
+    priv = this->private;
+
+    need_open_count = AFR_COUNT(local->need_open, priv->child_count);
+    if (!need_open_count) {
+        goto out;
+    }
+    gf_msg_debug(this->name, 0, "need open count: %d", need_open_count);
+    local->call_count = need_open_count;
+
+    for (i = 0; i < priv->child_count; i++) {
+        if (!local->need_open[i])
+            continue;
+
+        if (IA_IFDIR == local->fd->inode->ia_type) {
+            gf_msg_debug(this->name, 0, "opening fd for dir %s on subvolume %s",
+                         local->loc.path, priv->children[i]->name);
+            STACK_WIND_COOKIE(frame, afr_openfd_fix_open_cbk, (void *)(long)i,
+                              priv->children[i],
+                              priv->children[i]->fops->opendir, &local->loc,
+                              local->fd, NULL);
+        } else {
+            gf_msg_debug(this->name, 0,
+                         "opening fd for file %s on subvolume %s",
+                         local->loc.path, priv->children[i]->name);
+
+            STACK_WIND_COOKIE(
+                frame, afr_openfd_fix_open_cbk, (void *)(long)i,
+                priv->children[i], priv->children[i]->fops->open, &local->loc,
+                local->fd_ctx->flags & ~(O_CREAT | O_EXCL | O_TRUNC), local->fd,
+                NULL);
+        }
+        if (!--need_open_count)
+            break;
+    }
+    return 0;
+
+out:
+    AFR_STACK_DESTROY(frame);
+    return 0;
+}
+
+static int
+afr_is_reopen_allowed_cbk(call_frame_t *frame, void *cookie, xlator_t *this,
+                          int32_t op_ret, int32_t op_errno,
+                          struct gf_flock *lock, dict_t *xdata)
+{
+    afr_local_t *local = frame->local;
+    afr_private_t *priv = NULL;
+    int ret = -1;
+    int call_count = 0;
+    int i = (long)cookie;
+    int32_t is_reopen_allowed = -1;
+
+    priv = this->private;
+    local->replies[i].valid = 1;
+    local->replies[i].op_ret = op_ret;
+    local->replies[i].op_errno = op_errno;
+    if (op_ret != 0) {
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, AFR_MSG_LK_HEAL_DOM,
+               "Failed getlk for %s", uuid_utoa(local->fd->inode->gfid));
+    }
+
+    LOCK(&frame->lock);
+    {
+        if (local->is_reopen_allowed != 0) {
+            ret = dict_get_int32(xdata, "is-reopen-allowed-on-fd",
+                                 &is_reopen_allowed);
+            if (ret) {
+                gf_msg(this->name, GF_LOG_ERROR, -ret, AFR_MSG_DICT_GET_FAILED,
+                       "Failed to get whether reopen is allowed or not on fd "
+                       "for file %s on subvolume %s.",
+                       local->loc.path, priv->children[i]->name);
+            } else if (is_reopen_allowed == 0) {
+                local->is_reopen_allowed = 0;
+            } else if (is_reopen_allowed == 1) {
+                local->is_reopen_allowed = 1;
```

Comment:
Why not
```
else {
            local->is_reopen_allowed = is_reopen_allowed;
}
```

Reply:
If we do this, (1) we can end up reopening fds in cases where fd should not be reopened or (2) in cases where fd can be reopened but we might block that with this change. For example:
Case-1:
- local->is_reopen_allowed will be -1 initially.
- B1 needs reopen and gets FD_BAD from client_lk, 'local->is_reopen_allowed' will still be set to 2 with this change.
- B2 & B3 got disconnected & reconnected and gets FD_BAD from client_lk. This will set 'local->is_reopen_allowed' to FD_BAD/2.
- It will allow the reopen, which should not be allowed since we do not know whether there is another fd which is having lock.
For this we can check for local->is_reopen_allowed set to 1 @ line 400, which will handle this case.

case-2:
- There are no locks held
- B1 needs reopen, and get FD_BAD. local->is_reopen_allowed = 2 now
- B2 was always up and client_lk_cbk return FD_REOPEN_ALLOWED and sets local->is_reopen_allowed = 1
- B3 got disconnected & reconnected, hence will get FD_BAD and sets the same in local.
Now in this case ideally we should allow the reopen, but it will be blocked.

Instead of assigning this to local, we can simply ignore this, which will work in all the cases.

Reply:
I think you should add this as code comment too as to why it is done this way. is-reopen-allowed is a bad key name if it has 3 states. It is confusing :-( 
In that case we can have:
```
else if (is_reopen_allowed >= 0 && is_reopen_allowed < 2) {
            local->is_reopen_allowed = is_reopen_allowed;
}

Reply:
Done.